src/Lachain.Core/Network/BlockSynchronizer.cs

```diff
@@ -296,32 +298,31 @@ private void BlockSyncWorker()
                     const int maxBlocksToRequest = 10;
 
                     var peers = _peerHeights
-                        .Where(entry => entry.Value >= maxHeight)
-                        .Select(entry => entry.Key)
+                        .Where(entry => entry.Value > myHeight)
                         .OrderBy(_ => rnd.Next())
                         .Take(maxPeersToAsk)
                         .ToArray();
 
                     var leftBound = myHeight + 1;
-                    var rightBound = Math.Min(maxHeight, myHeight + maxBlocksToRequest);
-                    Logger.LogTrace($"Sending query for blocks [{leftBound}; {rightBound}] to {peers.Length} peers");
                     foreach (var peer in peers)
                     {
+                        var rightBound = Math.Min(peer.Value, myHeight + maxBlocksToRequest);
+                        Logger.LogTrace($"Sending query for blocks [{leftBound}; {rightBound}] to peer {peer.Key.ToHex()}");
                         _networkManager.SendTo(
-                            peer, _networkManager.MessageFactory.SyncBlocksRequest(leftBound, rightBound),
+                            peer.Key, _networkManager.MessageFactory.SyncBlocksRequest(leftBound, rightBound),
                             NetworkMessagePriority.PeerSyncMessage
                         );
                     }
 
-                    var waitStart = TimeUtils.CurrentTimeMillis();
-                    while (true)
+                    var waitForBlockExecution = 4000;
+                    bool gotResponse = true;
+                    lock (_peerHasBlocks)
                     {
-                        lock (_peerHasBlocks)
-                        {
-                            Monitor.Wait(_peerHasBlocks, TimeSpan.FromMilliseconds(1_000));
-                        }
-
-                        if (TimeUtils.CurrentTimeMillis() - waitStart > 5_000) break;
+                        gotResponse = Monitor.Wait(_peerHasBlocks, TimeSpan.FromMilliseconds(1_000));
+                    }
+                    if (gotResponse)
+                    {
+                        Thread.Sleep(waitForBlockExecution);
```

Comment:
Are we really need to use Sleep instead  of waiting for event?

Reply:
we wait 1 sec for reply, if we get no reply, we don't sleep, send another request, but if we get reply, then I think we should wait for some time for the incoming blocks to finish execution. I can try to handle it in such a way that we wait until execution of receiving blocks is finished and not use hardcoded value. Should I do it?

Reply:
ok